experimental/message/src/parser/mod.rs

```diff
@@ -0,0 +1,405 @@
+        assert_eq!(self.next(), Some(&b'n'));
+        let mut key = SmallVec::new();
+
+        self.skip_ws();
+
+        if self.next_if(b'*') {
+            key.push(ast::VariantKey::Asterisk);
+        }
+
+        self.skip_ws();
+
+        assert_eq!(self.next(), Some(&b'{'));
+
+        let pattern = self.parse_pattern()?;
+
+        Ok(ast::Variant { key, pattern })
+    }
+
+    fn parse_pattern(&mut self) -> ParserResult<ast::Pattern<S>> {
+        let mut start = self.ptr;
+        let mut body = SmallVec::new();
+        while let Some(b) = self.next() {
+            match b {
+                b'}' => {
+                    let end = self.ptr - 1;
+                    if start != end {
+                        body.push(ast::PatternElement::Text(self.source.slice(start..end)));
+                    }
+                    return Ok(ast::Pattern { body });
+                }
+                b'{' => {
+                    let end = self.ptr - 1;
+                    if start != end {
+                        body.push(ast::PatternElement::Text(self.source.slice(start..end)));
+                    }
+                    body.push(ast::PatternElement::Placeholder(self.parse_placeholder()?));
+                    start = self.ptr;
+                }
+                _ => {}
+            }
+        }
+        unreachable!()
+    }
+
+    fn parse_placeholder(&mut self) -> ParserResult<ast::Placeholder<S>> {
+        let placeholder = match get_current_byte!(self) {
+            Some(b'+') => {
+                self.ptr += 1;
+                let name = self.parse_name()?;
+                let options = SmallVec::new();
+                ast::Placeholder::Markup { name, options }
+            }
+            Some(b'-') => {
+                self.ptr += 1;
+                let name = self.parse_name()?;
+                ast::Placeholder::MarkupEnd { name }
+            }
+            Some(_) => {
+                let exp = self.parse_expression()?;
+                ast::Placeholder::Expression(exp)
+            }
+            None => {
+                unreachable!();
+            }
+        };
+        assert_eq!(self.next(), Some(&b'}'));
+        Ok(placeholder)
+    }
+
+    fn parse_expression(&mut self) -> ParserResult<ast::Expression<S>> {
+        let operand = self.parse_operand()?;
+        Ok(ast::Expression::Operand {
+            operand,
+            annotation: None,
+        })
+    }
+
+    fn parse_operand(&mut self) -> ParserResult<ast::Operand<S>> {
+        let op = match self.next() {
+            Some(b'$') => ast::Operand::Variable(self.parse_name()?),
+            Some(b'(') => ast::Operand::Literal(self.parse_literal()?),
+            _ => {
+                unreachable!()
+            }
+        };
+        Ok(op)
+    }
+
+    fn parse_name(&mut self) -> ParserResult<S> {
+        let start = self.ptr;
+        while let Some(b) = get_current_byte!(self) {
+            if b.is_ascii_alphabetic() {
```

Comment:
The actual `Name` spec is a bit different: https://github.com/unicode-org/message-format-wg/blob/develop/spec/syntax.md#names